apps/felles-auth/src/main/kotlin/no/nav/helsearbeidsgiver/felles/auth/AuthClient.kt

```diff
@@ -0,0 +1,93 @@
+import no.nav.helsearbeidsgiver.utils.log.sikkerLogger
+
+/** Krever `texasEnabled: true` i konfig. */
+class AuthClient {
+    private val sikkerLogger = sikkerLogger()
+    private val httpClient = createHttpClient()
+
+    private val tokenEndpoint = "NAIS_TOKEN_ENDPOINT".fromEnv()
+    private val tokenExchangeEndpoint = "NAIS_TOKEN_EXCHANGE_ENDPOINT".fromEnv()
+    private val tokenIntrospectionEndpoint = "NAIS_TOKEN_INTROSPECTION_ENDPOINT".fromEnv()
+
+    fun tokenGetter(
+        identityProvider: IdentityProvider,
+        target: String,
+    ): () -> String =
+        {
+            runBlocking {
+                sikkerLogger.info("Forsøker å hente token for '$target' fra '${identityProvider.verdi}' og endepunkt '$tokenEndpoint'.")
+                val tokenResponse = token(identityProvider, target)
+                sikkerLogger.info("Hentet token som utløper om ${tokenResponse.expiresInSeconds} sekunder.")
+                tokenResponse.accessToken
+            }
+        }
+
+    internal suspend fun token(
+        provider: IdentityProvider,
+        target: String,
+    ): TokenResponse =
+        try {
+            httpClient
+                .submitForm(
+                    url = tokenEndpoint,
+                    formParameters =
+                        parameters {
+                            identityProvider(provider)
+                            target(target)
+                        },
+                ).body()
+        } catch (e: ResponseException) {
+            onError(e.response)
+        }
+
+    internal suspend fun exchange(
+        provider: IdentityProvider,
+        target: String,
+        userToken: String,
+    ): TokenResponse =
+        try {
+            httpClient
+                .submitForm(
+                    url = tokenExchangeEndpoint,
+                    formParameters =
+                        parameters {
+                            identityProvider(provider)
+                            target(target)
+                            userToken(userToken)
+                        },
+                ).body()
+        } catch (e: ResponseException) {
+            onError(e.response)
+        }
+
+    internal suspend fun introspect(
+        provider: IdentityProvider,
+        accessToken: String,
+    ): TokenIntrospectionResponse =
+        httpClient
+            .submitForm(
+                url = tokenIntrospectionEndpoint,
+                formParameters =
+                    parameters {
+                        identityProvider(provider)
+                        token(accessToken)
+                    },
+            ).body()
+
+    private suspend fun onError(response: HttpResponse): Nothing {
+        val error = response.body<ErrorResponse>()
+        val msg = "Klarte ikke hente token. Feilet med status '${response.status}' og feilmelding '${error.errorDescription}'."
+
+        sikkerLogger.error(msg)
+        throw RuntimeException(msg)
```

Comment:
Bør man kaste et eget exception (ikke runtimeex) som evt fanges? Hvordan tenker du håndtering dersom kallet feiler?

Reply:
Hvis det feiler så tryner appen. Egentlig så tenker jeg det er greit, ettersom det sjeldent (aldri?) har skjedd med den nåværende løsningen, som oppfører seg likt. Hvis vi vil ha noe som skal fanges så tenker jeg vi likegodt kan throwe den originale `ResponseException`. Hva tenker du om det?

Reply:
Ja! (Til å kaste ResponseException) :)
Tenker at det kan være greit at appen ikke tryner, det bør vel f eks være bra å bare svare med en feilmelding og la klienten forsøke igjen etter en liten stund ved nettverksfeil o.l.

Reply:
Da endrer jeg exception.

Jeg var litt for kjapp for å melde tryning. Alle appene i Simba som bare driver med klientting har en feilhåndtering for ukjente exceptions, så de vil _ikke_ tryne, men heller sende en feilmelding på Kafka.

Reply:
Ok, det høres bedre ut!